server/task_executor_database_backup.go

```diff
@@ -88,18 +88,40 @@ func (*DatabaseBackupTaskExecutor) backupDatabase(ctx context.Context, server *S
 	}
 	defer driver.Close(ctx)
 
-	f, err := os.Create(filepath.Join(server.profile.DataDir, backup.Path))
+	backupFilePathLocal := filepath.Join(server.profile.DataDir, backup.Path)
+	backupFile, err := os.Create(backupFilePathLocal)
 	if err != nil {
 		return "", errors.Errorf("failed to open backup path: %s", backup.Path)
 	}
-	defer f.Close()
+	defer backupFile.Close()
 
-	payload, err := driver.Dump(ctx, databaseName, f, false /* schemaOnly */)
+	payload, err := driver.Dump(ctx, databaseName, backupFile, false /* schemaOnly */)
 	if err != nil {
 		return "", err
 	}
 
-	return payload, nil
+	switch backup.StorageBackend {
+	case api.BackupStorageBackendLocal:
+		return payload, nil
+	case api.BackupStorageBackendS3:
+		log.Debug("Uploading backup to s3 bucket.", zap.String("bucket", server.s3Client.GetBucket()), zap.String("path", backupFilePathLocal))
+		bucketFileToUpload, err := os.Open(backupFilePathLocal)
+		if err != nil {
+			return "", errors.Wrapf(err, "failed to open backup file %q for uploading to s3 bucket", backupFilePathLocal)
+		}
+		defer bucketFileToUpload.Close()
+		if _, err := server.s3Client.UploadObject(ctx, backup.Path, bucketFileToUpload); err != nil {
+			return "", errors.Wrapf(err, "failed to upload backup to AWS S3")
+		}
+		log.Debug("Successfully uploaded backup to s3 bucket.")
+		if err := os.Remove(backupFilePathLocal); err != nil {
```

Comment:
If it's cloud storage should we call os.Remove in a defer method?

Reply:
It's better to delete the local file just after uploading is successful.

Reply:
What's the use of the local file if the upload fails? Or put it another way, how to deal with it?

Reply:
Oh, I see your point. We should always delete it.

Reply:
Fixed.